### ISSUE TITLE
Procedural request[s]

### ISSUE THREAD

**Issue by dot-asm** [openssl/openssl#3963]
There are different ways to address feedback provided in the course of handling merge request. One is to push incremental changes, another is to maintain single [or same amount of] commit[s] throughout lifetime of merge request by force-pushing. Maintaining single [or same amount of] commit[s] past point of first approval makes reviewer's life harder than necessary. Well, formally speaking it makes life harder even prior first approval. Because maintaining single commit leaves no easy way for reviewer to "see" that there were no other changes besides those discussed. But approval is a check-point where reviewer has to make an *overall* assessment in either case, so maintaining single commit prior first approval *might* be considered acceptable. "Might" in previous sentence refers to the fact that life is compromise, and just formulated suggestion is not universal. In this case compromise is mostly about changes' complexity and size. I mean for big enough changes it might be *in*appropriate to maintain single commit at *any* step. [And reverse it true, from small, trivial changes it might be perfectly appropriate to maintain single commit all along.]

To summarize. Suggestion is to ask contributors ~is~ to abstain from maintaining single [or same amount of] commit[s] at least past first approval, ~but~ and consider incremental changes even prior that point. This doesn't preclude possibility that contributor can be asked to squash changes. On the other hand even though it's not explicitly proclaimed, the project reserves some freedom of editorial right to squash commit upon commit to repository. For this reason commits tagged with say "[to be squashed]" should be welcomed.

**Comment by levitte** [comment]
I would further like to suggest that squashing all commits into one big upon final merge should be frowned upon.  The '[to be squashed]' commits should, of course, be squashed with the commit they refer to, but otherwise, squashing should at least be done with care.  The exception is, of course, if the overall change is small.

For example, I frown upon putting source style changes and other more substantial changes into a single commit in a PR, and I think they should be separate in the final merge as well.  Otherwise, looking back at history becomes just as hellish as a PR that holds everything in one commit.

**Comment by dot-asm** [comment]
> For example, I frown upon putting source style changes and other more substantial changes into a single commit in a PR, and I think they should be separate in the final merge as well.

One probably can't disagree as long as we are talking about styling of existing code, existing at the moment of request(\*). But I'd argue that styling fixes to code introduced in given request are better squashed. But let's not forget that suggestion is rather about specific point in the review process. I for one see no reason to be rigid about what happens prior that point, reviewer is free to ask for whatever action that is deemed appropriate, to split or squash, or insist on incremental squashables. It's rather about what happens past the point in question.

(\*) At the same time if styling modification is just one and couple of lines away, it's probably ok.

**Comment by kaduk** [comment]
I generally agree with @dot-asm 's sentiment here in terms of supporting the reviewer.  However, things become awkward when a pull request remains around long enough that merge conflicts with the target branch develop -- in such cases, the CI systems refuse to even attempt to build the pull request when changes are made.  Or do we want people to merge master into their pull request to resolve the conflicts and the committer to rebase prior to actually pushing to the main repo?

**Comment by paulidale** [comment]
Sounds reasonable.

**Comment by bernd-edlinger** [comment]
I also try to use a good commit message and sometimes getting smarter
after the fist review loop...

**Comment by dot-asm** [comment]
Formally speaking one can can probably take it one step further and even argue that squashing (again, past point of first review) is actually reviewers' *responsibility*. However, it doesn't actually mean that reviewer has to do the actual squashing, one can still turn to the contributor. But then in majority of such cases it would be more than appropriate for reviewer to pull a local copy *prior* asking contributor to do the squashing, so that outcome can be double-checked (as suggested by @kaduk in #3992). As failure to do so would formally require reviewer to restart review from scratch. But one way or another it's probably appropriate to say that squashing reviewers' responsibility and consequently choice of actual method [which automatically would mean that contributor is not free to take the initiative to squash].

**Comment by bernd-edlinger** [comment]
When I compare the review process here with the review process for GCC patches,
there are some differences.

For instance every GCC patch is first prepared and sent to the gcc-patches mailing list.
It is always sent together with the proposed ChangeLog entry.
The reviewer either approves or requests changes on the patch itself or the ChangeLog entry.
Then the author updates his patch and sends it again to the gcc-patches ML.
In effect it is possible to find every version of each patch together with the review comments
and compare what was checked in to what was finally approved.

OTOH when I have a closed PR and want to compare the approved version with the
finally checked in version, all I can see is the review comments, because once the author
removes the branch on his repository, all history is gone.  And the author has less control
over the ChangeLog entry, because after the squashing the committer creates an completely
new ChangeLog entry out of the different commits.

**Comment by dot-asm** [comment]
Well, github is not really a review system, and my suggestion is admittedly just an attempt to make best of situation.

**Comment by bernd-edlinger** [comment]
Hmm, yes, interesting, usually each commit, is recoded in the git reflog, and
the reflog gets garbage collected after 90 days, but I see squashed commits
much older than that. So It looks like github does not garbage collect unreachable
commits.  I wonder if we can rely on this "feature", and what happens if the author's
github account is deleted.  Why is the "View Changes" not working when the commit
is still in the github tree?

**Comment by dot-asm** [comment]
> git reflog

Isn't reflog a local thing? I mean let's say you look at merge request page for first time and want to double-check outcome of squash operation. What do you pull to recover pre-squash hash? There is no reflog that can help you...

**Comment by kaduk** [comment]
Yes, reflog is a local thing.
The only ways I know of to get a commit hash for subsequent examination is if someone made an inline comment on that commit, or maybe if github had generated an email when a new commit was added to the PR (which isn't helpful in the general case).  And I don't understand how github decides when/whether to send emails for new commits on pull requests...

**Comment by bernd-edlinger** [comment]
This is the info I just received from github support:

> Hey Bernd,
> 
> We normally wouldn't garbage collect a repository unless its requested by the repository owners.
> 
> This means that commits may hang around but we cannot guarantee that they'll always be there.
> 
> Any commit thats referenced in another branch or fork will remain though. Even if you delete it from one branch it will still be accessible via any other ref it's still a part of.
> 
> Cheers
> Stacey
> 
>     Hi,
> 
>     could you chime in on openssl Issue #3963 ?
>     It looks like that reviewed changes are often available even after a force
>     push or author deleting the branch.
>     Just the "View Changes" button does not work, but it turns out
>     that old commits are often available quite long after they are usually
>     garbage collected.
>     I wonder if we can rely on that or if the commits will eventually be
>     deleted?
>

**Comment by dot-asm** [comment]
This sounds that mentioning pre-squash commit in conversation (so that there is an URL) is not guaranteed to work. It's likely to work, but not *guaranteed*, and hence should not be relied upon. So that explicit pre-squash pull would be preferred option? Anyway, all this is relevant only if we actually agree on the premise, namely that squashing [specifically past point of first approval] is reviewers' responsibility (and not contributor's initiative). [Once again, "responsibility" doesn't necessarily mean that reviewer is obliged to do the squashing, only to ensure that outcome is consistent, and doing it yourself is simply one way to achieve the goal.]

There also was question about how to handle re-bases. In the context of this discussion more specifically what if there is *approved* request, but which has conflicts, and question is if commiter is free to resolve conflicts upon merge to main repository. I'd be inclined to say that this is error-prone procedure that might be inappropriate for this project. Or in other words I for one would argue in favour of handling re-base *exclusively* within merge request irregardless its approval status. So that re-base would always be contributor's responsibility. I would even take it one step further and say that re-base effectively voids [all] approval[s]. As compromise corner case, one can probably argue that if re-base outcome is a squashable that involves for example some ordinals re-numbering[?], then it doesn't have to void the approval. But I have feeling that there already is kind of implicit consensus at least on the part that re-bases are to be handled at review time and never at merge time, i.e. merge to main repo has to be conflict-free.

**Comment by dot-asm** [comment]
Another procedural request. I'd like to suggest to make it request originators' responsibility to explicitly comment on red-crosses from CI. Effectively it means that request would be less likely to be reviewed if there is red cross from either CI and no note from contributor.
